src/Poisson2DSparseMatrix.cpp

```diff
@@ -6,16 +6,191 @@
 #include "MeshManager.hpp"
 #include "VandermondeBuilders.hpp"
 #include "BCtypes.hpp"
+#include "BlitzHelpers.hpp"
 #include "Types.hpp"
 
 using blitz::Range;
 using boost::python::numpy::ndarray;
 using boost::python::numpy::zeros;
 using boost::python::numpy::dtype;
+using boost::python::dict;
 
 namespace blitzdg{
     Poisson2DSparseMatrix::Poisson2DSparseMatrix(DGContext2D& dg, MeshManager& mshManager) {
+        const index_vector_type& bcType = mshManager.get_BCType();
+        buildPoissonOperator(dg, mshManager, bcType);
+    }
+
+    void Poisson2DSparseMatrix::buildBcRhs(DGContext2D& dg, const MeshManager& mshManager, const real_matrix_type& ubc, const real_matrix_type& qbc, const index_vector_type& bcType) {
+        const index_type Np = dg.numLocalPoints();
+        const index_type K = dg.numElements();
+        const index_type Nfaces = dg.numFaces();
+        const index_type Nfp = dg.numFacePoints();
+
+        real_matrix_type bcRhs(Np, K);
+        bcRhs = 0.0;
+
+        const real_matrix_type& Dr = dg.Dr();
+        const real_matrix_type& Ds = dg.Ds();
+
+        const real_matrix_type& rx = dg.rx();
+        const real_matrix_type& ry = dg.ry();
+        const real_matrix_type& sx = dg.sx();
+        const real_matrix_type& sy = dg.sy();
+        const real_matrix_type& J  = dg.jacobian();
+        const real_matrix_type& Fscale = dg.fscale();
+
+        const index_vector_type& EToE = mshManager.get_EToE();
+        const index_vector_type& EToF = mshManager.get_EToF();
+
+        blitz::firstIndex ii;
+        blitz::secondIndex jj;
+        blitz::thirdIndex kk;
+
+        const index_type N = dg.order();
+
+        const index_vector_type& vmapM = dg.vmapM();
+        const index_vector_type& vmapP = dg.vmapP();
+
+        const index_matrix_type& Fmask = dg.fmask();
```

Comment:
It looks like `Fmask` is accessed by its columns, so you have it stored in column-major order for cache locality. Not sure if that matters here, just a thought.

Reply:
Fmask is pretty small so I doubt that it matters in this case. I will switch to column-major and see if anything breaks though.